nixos/mpd: apply RFC 42

doronbehar opened the pull request:
## Description of changes


## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

doronbehar committed:
nixos/mpd: move user, group & dataDir upwards (preparation commit)

doronbehar committed:
nixos/mpd: rename some settings to fit upstream's names

doronbehar committed:
nixos/mpd: small indentation fix

doronbehar committed:
nixos/mpd: small db_file type fix

doronbehar committed:
nixos/mpd: always set all options explicitly

doronbehar committed:
nixos/mpd: add openFirewall option

doronbehar committed:
nixos/mpd: apply RFC 42

Also use `config.services.mpd.settings` directly in the documentation of
mpdscribble.

doronbehar committed:
nixosTests.mpd: squash defaultCfg into defaultMpdCfg

doronbehar committed:
nixosTests.mpd: don't explicitly set default musicDirectory

doronbehar committed:
nixosTests.mpd: refresh example music track

The previous track was 404 probably due to changes in the website.

doronbehar committed:
nixosTests.mpd: put mpd.conf files in $out

doronbehar committed:
nixosTests.mpd: add doronbehar as maintainer

doronbehar committed:
mpd: add doronbehar as maintainer

doronbehar committed:
mpd: use lib.meson* functions

emmanuelrosa reviewed (approved):
Looks good to me.
Awesome work, @doronbehar.

doronbehar committed:
mpd: use --replace-fail in darwin postPatch substitution

h7x4 commented on nixos/modules/services/audio/mpd.nix:
```diff
@@ -126,49 +117,136 @@ in
+            ];
+          options = {
+            music_directory = lib.mkOption {
+              type = with lib.types; either path (strMatching "(http|https|nfs|smb)://.+");
```
I believe what schemas are valid depends on what storage plugins you have activated. There is at least `udisks://` as well. Maybe it would be better to just ensure that it looks like a URI somehow, rather than hardcode the schemas

h7x4 commented on nixos/modules/services/audio/mpd.nix:
```diff
@@ -12,46 +12,60 @@ let
-    ${cfg.extraConfig}
-  '';
+    ];
+  mpdConf = pkgs.writeText "mpd.conf" (
```
It would be kinda cool if we could validate this file. Maybe it's possible to run mpd with the config file through gnu expect in the nix sandbox during a check phase, and see if it throws a config-related error.

doronbehar replied:
> ```diff
> +            attrsOf (oneOf [
>                atomType
> -              (attrsOf (listOf (attrsOf atomType)))
> -            ];
> +              (listOf (attrsOf atomType))
> +            ]);
>  
> ```

Thanks!

> ```diff
> +            ++ lib.optionals (cfg.dataDir == "/var/lib/${name}") [ name ]
> ```

Doesn't it mean you still depend on a `dataDir` configuration value? I thought you wish to get rid of it in favor of only `state_file` and `sticker_file`.

> ```diff
> +            expectScript = ''
> +              spawn mpd --no-daemon "$env(configFilePath)"
> +              expect {
> +                "exception: Error in \"$env(configFilePath)\"" {
> +                  puts "Config file invalid\n"
> +                  exit 1
> +                }
> +                "exception:" {
> +                  exit 0
> +                }
> +              }
> +            '';
> +
> +            configFile = configContent;
> +            passAsFile = [
> +              "expectScript"
> +              "configFile"
> +            ];
> +
> +            doCheck = true;
> +            nativeCheckInputs = [
> +              pkgs.expect
> +              pkgs.mpd
> +            ];
> +            checkPhase = ''
> +              runHook preCheckPhase
> +              expect -f "$expectScriptPath"
> +              runHook postCheckPhase
> +            '';
> +
> +            installPhase = ''
> +              runHook preCheckPhase
> +              install -Dm444 "$configFilePath" "$out"
> +              runHook postCheckPhase
> +            '';
> ```

Nice! I used your `expectScript` and used `pkgs.writeTextFile` instead of `pkgs.writeText`, which allows specifying `checkPhase` and the other `derivationArgs`.

> ```diff
> +            renderNonBlock = k: v: ''${k} "${if builtins.isBool v then lib.boolToYesNo v else toString v}"'';
> +            renderBlock = n: v: ''
> +              ${n} {
> +              ${lib.concatMapStringsSep "\n" (l: "    " + l) (lib.mapAttrsToList renderNonBlock v)}
> +              }
> +            '';
> ```

This seems perfectly fine, but I hope you'd forgive me for not incorporating these signatures.

> ```diff
> +      decoder = lib.mkIf cfg.fluidsynth [
> +        {
> +          plugin = "fluidsynth";
> +          soundfont = "${pkgs.soundfont-fluid}/share/soundfonts/FluidR3_GM2-2.sf2";
> +        }
> +      ];
> +    };
> ```

Are you sure this works better then my code? It seems that defining multiple decoders along with `services.mpd.fluidsynth`? Why not handle `cfg.fluidsynth` before processing the blocks?

> ```diff
> -          autio_output = [
> +          audio_output = [
> ```

Thanks!

> * Move the `atomType` down to where it is used

(EDITED) I _have_ incorporated this change, thanks too!

> * Move the file config file derivation down to around where it is used.

(EDITED) I haven't incorporated _this_ change, to make the diff as small as possible.

h7x4 commented on nixos/modules/services/audio/mpd.nix:
```diff
@@ -126,49 +117,136 @@ in
-        description = "User account under which MPD runs.";
+      openFirewall = lib.mkOption {
+        type = lib.types.bool;
+        default = lib.elem cfg.settings.bind_to_address [
```
I'm not entirely sure I'm sold on this idea. I think there is some value in having the user explicitly say that the service should be let through the firewall.

Considering the default address is `127.0.0.1` and that most machines running mpd likely will be behind a home router, it's not a big deal. This is probably not the first module to do this either. But I still find it a bit spooky that I could accidentally open the firewall through misconfiguration seemingly unrelated to the firewall.

doronbehar replied:
> But I still find it a bit spooky that I could accidentally open the firewall through misconfiguration seemingly unrelated to the firewall.

I agree, and I have 2 suggestions:

- Warn about this in the `bind_to_address` documentation.
- Open the firewall only if `bind_to_address == "any"`.

What do you think?

doronbehar committed:
nixos/mpd: allow any URI schema in music_directory

There are many more URI schemas possible, and it depends on storage
plugins activated etc.

doronbehar commented on nixos/modules/services/audio/mpd.nix:
```diff
@@ -126,49 +117,136 @@ in
+          "0.0.0.0"
+        ];
+        defaultText = lib.literalExpression ''
+          elem config.services.mpd.settings.bind_to_address ["any" "0.0.0.0"]
```
I used `builtins.elem` as `lib.elem` is just aliased to it.

doronbehar commented on nixos/modules/services/audio/mpd.nix:
```diff
@@ -126,49 +117,136 @@ in
-        type = lib.types.nullOr lib.types.str;
-        default = "${cfg.dataDir}/tag_cache";
-        defaultText = lib.literalExpression ''"''${dataDir}/tag_cache"'';
+      blocks = lib.mkOption {
```
> Maybe it would be simpler and more in line with the RFC42 idea to just allow `listOf (attrsOf <atomType>)` in the `freeformType`, and then the key for the list would be the `mpd_key`?

That's a great idea! I'm working on it now.

doronbehar committed:
nixos/mpd: use pkgs.writeTextFile, with a checkPhase

doronbehar committed:
nixos/mpd: small formatting to mpdConf

doronbehar committed:
nixos/mpd: nixfmt

tobim reviewed (commented):
While I unfortunately don't have capacity to review such a big change at the moment I still wanted to say I wholeheartedly support this effort! ❤️

tobim reviewed (approved):
Tested this branch with this config in conjunction with with `snapserver` and it works fine:
```nix
  services.mpd = {
    enable = true;
    musicDirectory = "/media/music";
    network.listenAddress = "any";
    openFirewall = true;
    settings.audio_output = [
      {
        type = "fifo";
        name = "snapcast";
        path = "/run/snapserver/mpd";
        format = "48000:16:2";
        mixer_type = "software";
      }
    ];
  };
```

doronbehar commented on nixos/modules/services/audio/mpd.nix:
```diff
@@ -126,48 +142,144 @@ in
+        default = builtins.elem cfg.settings.bind_to_address [
+          "any"
+          "0.0.0.0"
+        ];
```
> Maybe this is a bit much magic, defaults depending on other values can be surprising. I'm also not aware of any other modules that do something similar.

Yea I talked about this with @h7x4 and I tried to satisfy this comment in the following note for `bind_to_address`'s documentation:

```md
NOTE: Using `any` or `0.0.0.0` will result in the configured port to
be open in the firewall.
```

I won't mind too much to default it to `false`, but I just wish to verify first you have noticed the above.

tobim replied:
I hadn't seen it because it was collapsed in my view.
My personal preference would be to default `openFirewall` to false and print a warning if the value of `bind_to_address` is anything other than `"127.0.0.1"` or `"localhost"` AND `openFirewall` is `false`.

misuzu commented on nixos/modules/services/audio/mpd.nix:
```diff
@@ -235,8 +338,47 @@ in
+          && !cfg.openFirewall
+        )
+        "Using '${cfg.bind_to_address}' as services.mpd.settings.bind_to_address without enabling services.mpd.openFirewall, might prevent you from accessing MPD from other clients.";
+
```
I'm using other means to access mpd, how can I remove this warning without setting `services.mpd.openFirewall = true;`?
```
evaluation warning: Using 'any' as services.mpd.settings.bind_to_address without enabling services.mpd.openFirewall, might prevent you from accessing MPD from other clients.
```

doronbehar replied:
> I'm using other means to access mpd, how can I remove this warning without setting `services.mpd.openFirewall = true;`?
> 
> ```
> evaluation warning: Using 'any' as services.mpd.settings.bind_to_address without enabling services.mpd.openFirewall, might prevent you from accessing MPD from other clients.
> ```

How do you access `mpd`?

misuzu replied:
Only over tailscale:
```
  networking.firewall.trustedInterfaces = [
    config.services.tailscale.interfaceName
  ];
```

misuzu replied:
> If you could share more details I might be able to help. I'm not familiar with tailscale and from skimming over its readme it is not trivially clear to me.

It's a VPN. I don't want to expose mpd to non-tailscale networks.


> BTW are you absolutely sure that using `127.0.0.1` will block you from access the mpd server?

I won't be able to access it from other devices if it listens on localhost.

misuzu replied:
I'm using `networking.firewall.trustedInterfaces` to allow all ports to be available on my tailscale network. I just don't want to see any warnings. This is the only module that I use which has this issue.
